File: content/14-2d-dynamic-programming/longest_common_subsequence.py
"""
Problem: Longest Common Subsequence
LeetCode: #1143 (Medium) | NeetCode 150 | Blind 75

Given two strings text1 and text2, return the length of their longest common subsequence.

Time Complexity: O(M * N)
Space Complexity: O(M * N)
"""

def longest_common_subsequence(text1: str, text2: str) -> int:
    m, n = len(text1), len(text2)
    dp = [[0] * (n + 1) for _ in range(m + 1)]
    for i in range(m - 1, -1, -1):
        for j in range(n - 1, -1, -1):
            if text1[i] == text2[j]:
                dp[i][j] = 1 + dp[i + 1][j + 1]
            else:
                dp[i][j] = max(dp[i + 1][j], dp[i][j + 1])
    return dp[0][0]

if __name__ == "__main__":
    assert longest_common_subsequence("abcde", "ace") == 3
    assert longest_common_subsequence("abc", "abc") == 3
    assert longest_common_subsequence("abc", "def") == 0
    print("All Python tests passed for Longest Common Subsequence!")

File: content/14-2d-dynamic-programming/longest_common_subsequence.js
/**
 * Problem: Longest Common Subsequence
 * LeetCode: #1143 (Medium) | NeetCode 150 | Blind 75
 * 
 * Time Complexity: O(M * N)
 * Space Complexity: O(M * N)
 */

function longestCommonSubsequence(text1, text2) {
    const m = text1.length, n = text2.length;
    const dp = Array.from({ length: m + 1 }, () => new Array(n + 1).fill(0));
    for (let i = m - 1; i >= 0; i--) {
        for (let j = n - 1; j >= 0; j--) {
            if (text1[i] === text2[j]) {
                dp[i][j] = 1 + dp[i + 1][j + 1];
            } else {
                dp[i][j] = Math.max(dp[i + 1][j], dp[i][j + 1]);
            }
        }
    }
    return dp[0][0];
}

if (require.main === module) {
    console.assert(longestCommonSubsequence("abcde", "ace") === 3, "Test 1 Failed");
    console.assert(longestCommonSubsequence("abc", "abc") === 3, "Test 2 Failed");
    console.assert(longestCommonSubsequence("abc", "def") === 0, "Test 3 Failed");
    console.log("All JavaScript tests passed for Longest Common Subsequence!");
}
